File: apina-core/src/main/java/fi/evident/apina/model/ApiDefinition.java
package fi.evident.apina.model;

import fi.evident.apina.model.type.ApiArrayType;
import fi.evident.apina.model.type.ApiClassType;
import fi.evident.apina.model.type.ApiType;
import fi.evident.apina.model.type.ApiTypeName;

import java.util.*;
import java.util.stream.Stream;

import static fi.evident.apina.utils.CollectionUtils.concat;
import static fi.evident.apina.utils.CollectionUtils.optionalToStream;
import static java.util.Collections.unmodifiableCollection;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toSet;

/**
 * Represents the whole API of the program: i.e. all {@link EndpointGroup}s.
 */
public final class ApiDefinition {

    private final Collection<EndpointGroup> endpointGroups = new ArrayList<>();
    private final Map<ApiTypeName, ClassDefinition> classDefinitions = new TreeMap<>();
    private final Map<ApiTypeName, EnumDefinition> enumDefinitions = new TreeMap<>();
    private final Set<ApiTypeName> blackBoxTypes = new LinkedHashSet<>();

    public Collection<EndpointGroup> getEndpointGroups() {
        return unmodifiableCollection(endpointGroups);
    }

    public void addEndpointGroups(EndpointGroup group) {
        endpointGroups.add(requireNonNull(group));
    }

    public boolean containsType(ApiTypeName typeName) {
        return classDefinitions.containsKey(typeName) || enumDefinitions.containsKey(typeName);
    }

    public void addClassDefinition(ClassDefinition classDefinition) {
        verifyTypeDoesNotExist(classDefinition.getType());

        classDefinitions.put(classDefinition.getType(), classDefinition);
    }

    public Collection<ClassDefinition> getClassDefinitions() {
        return unmodifiableCollection(classDefinitions.values());
    }

    public void addEnumDefinition(EnumDefinition enumDefinition) {
        verifyTypeDoesNotExist(enumDefinition.getType());
        enumDefinitions.put(enumDefinition.getType(), enumDefinition);
    }

    private void verifyTypeDoesNotExist(ApiTypeName type) {
        if (containsType(type))
            throw new IllegalArgumentException("tried to add type-definition twice: " + type);
    }

    public Collection<EnumDefinition> getEnumDefinitions() {
        return unmodifiableCollection(enumDefinitions.values());
    }

    public int getEndpointGroupCount() {
        return endpointGroups.size();
    }

    public int getEndpointCount() {
        return endpointGroups.stream()
                .mapToInt(EndpointGroup::getEndpointCount)
                .sum();
    }

    public int getClassDefinitionCount() {
        return classDefinitions.size();
    }

    public int getEnumDefinitionCount() {
        return enumDefinitions.size();
    }

    public Collection<ApiTypeName> getAllBlackBoxClasses() {
        return concat(blackBoxTypes, getUnknownTypeReferences());
    }

    /**
     * Returns all class types that refer to unknown classes.
     */
    public Set<ApiTypeName> getUnknownTypeReferences() {
        Stream<ApiType> resultTypes = endpointGroups.stream()
                .flatMap(e -> e.getEndpoints().stream())
                .flatMap(e -> optionalToStream(e.getResponseBody()));

        Stream<ApiType> propertyTypes = classDefinitions.values().stream()
                .flatMap(c -> c.getProperties().stream())
                .map(PropertyDefinition::getType);

        return Stream.concat(resultTypes, propertyTypes)
                .flatMap(ApiDefinition::referencedClassTypes)
                .filter(c -> !containsType(c))
                .collect(toSet());
    }

    private static Stream<ApiTypeName> referencedClassTypes(ApiType type) {
        if (type instanceof ApiClassType) {
            return Stream.of(((ApiClassType) type).getName());
        } else if (type instanceof ApiArrayType) {
            return referencedClassTypes(((ApiArrayType) type).getElementType());
        } else {
            return Stream.empty();
        }
    }

    public void addBlackBox(ApiTypeName type) {
        blackBoxTypes.add(requireNonNull(type));
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/model/ApiDefinition.kt
package fi.evident.apina.model

import fi.evident.apina.model.type.ApiType
import fi.evident.apina.model.type.ApiTypeName
import java.util.*

/**
 * Represents the whole API of the program: i.e. all [EndpointGroup]s.
 */
class ApiDefinition {

    private val _endpointGroups = ArrayList<EndpointGroup>()
    private val _classDefinitions = TreeMap<ApiTypeName, ClassDefinition>()
    private val _enumDefinitions = TreeMap<ApiTypeName, EnumDefinition>()
    private val _discriminatedUnionDefinitions = TreeMap<ApiTypeName, DiscriminatedUnionDefinition>()
    private val blackBoxTypes = LinkedHashSet<ApiTypeName>()

    val endpointGroups: Collection<EndpointGroup>
        get() = _endpointGroups.sortedBy { it.name }

    val classDefinitions: Collection<ClassDefinition>
        get() = _classDefinitions.values

    val enumDefinitions: Collection<EnumDefinition>
        get() = _enumDefinitions.values

    val discriminatedUnionDefinitions: Collection<DiscriminatedUnionDefinition>
        get() = _discriminatedUnionDefinitions.values

    fun addEndpointGroups(group: EndpointGroup) {
        _endpointGroups += group
    }

    fun containsType(typeName: ApiTypeName) =
        typeName in _classDefinitions || typeName in _enumDefinitions || typeName in _discriminatedUnionDefinitions

    fun addClassDefinition(classDefinition: ClassDefinition) {
        verifyTypeDoesNotExist(classDefinition.type)

        _classDefinitions[classDefinition.type] = classDefinition
    }

    fun addEnumDefinition(enumDefinition: EnumDefinition) {
        verifyTypeDoesNotExist(enumDefinition.type)

        _enumDefinitions[enumDefinition.type] = enumDefinition
    }

    fun addDiscriminatedUnion(definition: DiscriminatedUnionDefinition) {
        verifyTypeDoesNotExist(definition.type)

        _discriminatedUnionDefinitions[definition.type] = definition
    }

    fun addBlackBox(type: ApiTypeName) {
        blackBoxTypes += type
    }

    private fun verifyTypeDoesNotExist(type: ApiTypeName) {
        if (containsType(type))
            throw IllegalArgumentException("tried to add type-definition twice: $type")
    }

    val endpointGroupCount: Int
        get() = _endpointGroups.size

    val endpointCount: Int
        get() = _endpointGroups.sumBy { it.endpointCount }

    val classDefinitionCount: Int
        get() = _classDefinitions.size

    val enumDefinitionCount: Int
        get() = _enumDefinitions.size

    val allBlackBoxClasses: Collection<ApiTypeName>
        get() = (blackBoxTypes + unknownTypeReferences).sorted()

    /**
     * Returns all class types that refer to unknown classes.
     */
    val unknownTypeReferences: Set<ApiTypeName>
        get() {
            val resultTypes = _endpointGroups.asSequence()
                .flatMap { it.endpoints.asSequence() }
                .mapNotNull { it.responseBody }

            val propertyTypes = _classDefinitions.values.asSequence()
                .flatMap { it.properties.asSequence() }
                .map { it.type }

            return (resultTypes + propertyTypes)
                .flatMap { referencedClassTypes(it) }
                .filterNot { containsType(it) }
                .toSet()
        }

    private fun referencedClassTypes(type: ApiType): Sequence<ApiTypeName> = when (type) {
        is ApiType.Class -> sequenceOf(type.name)
        is ApiType.Array -> referencedClassTypes(type.elementType)
        else -> emptySequence()
    }
}
